astropy/io/fits/column.py

```diff
@@ -1389,19 +1389,21 @@ def __getattr__(self, name):
 
         Implements for example self.units, self.formats, etc.
         """
-
         cname = name[:-1]
         if cname in KEYWORD_ATTRIBUTES and name[-1] == 's':
             attr = []
-            for col in self:
+            for col in self.columns:
                 val = getattr(col, cname)
-                if val is not None:
-                    attr.append(val)
-                else:
-                    attr.append('')
+                attr.append(val if val is not None else '')
             return attr
         raise AttributeError(name)
 
+    def _update_attrs(self, name=None):
+        """Get or update local lists of column attributes"""
+        attrs = [name] if name is not None else KEYWORD_ATTRIBUTES
+        for name in attrs:
+            setattr(self, name + 's', self.__getattr__(name + 's'))
```

Comment:
Yep, this is actually to avoid the many calls to `Coldefs.__getattr__`. Sorry took me some time to write this answer but I'm just back from vacation, time to reboot the brain ;)

So, just a quick explanation/example for @astrofrog and others: for one column attribute, say `format`, `Coldefs` has a `formats` attribute (with the `s`) which calls `__getattr__`, which build a list of `Column.format` for all columns. And this applies for all attributes listed in `KEYWORD_ATTRIBUTES`. 
With the current implementation (without this PR), `__getattr__` is called at least once for each column and attribute (from another comment earlier: with a table with 720 columns, `__getattr__` is called 722 times for 'format' and 1441 times for 'name' ...).

Also I agree with you @embray, it is making the code even more complicated, but it is a quite easy way to speed up things for now. And there are not so many required calls to `_update_attrs`. (And I guess this code is quite stable !).
I can also try to take another look if you think there is a better way (if it is not too difficult to find all uses of `cols.names[i]`).